File: packages/kdna-core/src/signature.js
/**
 * KDNA asset signatures — RFC-0021 M1 (`kdsig.ed25519`).
 *
 * A `.kdna` container carries an optional top-level `signature.kdsig` entry:
 * a JSON signature bundle whose Ed25519 signature covers a deterministic
 * signing payload derived from the asset's canonical content digest
 * (docs/CANONICALIZATION.md). Verification is offline and fail-closed: any
 * malformed, unsupported, or unverifiable bundle rejects the asset.
 *
 * A signature proves integrity and provenance only. It never proves
 * expertise, truthfulness, safety, or fitness for purpose.
 */

'use strict';

const crypto = require('crypto');

const KDSIG_PROFILE = 'kdsig.ed25519';
const KDSIG_PROFILE_VERSION = '0.1.0';
const KDSIG_ALGORITHM = 'ed25519';
const SIGNATURE_ENTRY_NAME = 'signature.kdsig';

const RAW_PUBLIC_KEY_HEX_LENGTH = 64;
const RAW_SIGNATURE_HEX_LENGTH = 128;
const SHA256_HEX_LENGTH = 64;

// RFC 8410 DER prefixes for raw Ed25519 key material.
const PKCS8_ED25519_PREFIX = Buffer.from('302e020100300506032b657004220420', 'hex');
const SPKI_ED25519_PREFIX = Buffer.from('302a300506032b6570032100', 'hex');

const BUNDLE_FIELDS = Object.freeze([
  'algorithm',
  'content_digest',
  'profile',
  'profile_version',
  'public_key',
  'signature',
]);

function sha256Hex(buf) {
  return crypto.createHash('sha256').update(buf).digest('hex');
}

function signatureError(code, message) {
  const error = new Error(message);
  error.code = code;
  return error;
}

function isLowercaseHex(value, length) {
  return typeof value === 'string' && value.length === length && /^[0-9a-f]+$/.test(value);
}

function assertContentDigestShape(value) {
  if (
    typeof value !== 'string'
    || !value.startsWith('sha256:')
    || !isLowercaseHex(value.slice('sha256:'.length), SHA256_HEX_LENGTH)
  ) {
    throw signatureError(
      'KDNA_INTEGRITY_SIGNATURE_FAILED',
      `signature bundle content_digest must be "sha256:<64 lowercase hex>", got ${JSON.stringify(value)}`,
    );
  }
}

/**
 * Build the exact bytes that Ed25519 signs. The payload is domain-separated by
 * the profile coordinate and binds the asset's canonical content digest, so a
 * signature can never be replayed across profiles, versions, or assets.
 */
function buildSigningPayload(contentDigest) {
  assertContentDigestShape(contentDigest);
  return Buffer.from(`${KDSIG_PROFILE}:${KDSIG_PROFILE_VERSION}:${contentDigest}`, 'utf8');
}

function privateKeyObjectFromSeed(seedHex) {
  if (!isLowercaseHex(seedHex, RAW_PUBLIC_KEY_HEX_LENGTH)) {
    throw signatureError(
      'KDNA_SIGNATURE_KEY_INVALID',
      'Ed25519 private key must be a 32-byte seed encoded as 64 lowercase hex characters',
    );
  }
  const der = Buffer.concat([PKCS8_ED25519_PREFIX, Buffer.from(seedHex, 'hex')]);
  try {
    return crypto.createPrivateKey({ key: der, format: 'der', type: 'pkcs8' });
  } catch (e) {
    throw signatureError('KDNA_SIGNATURE_KEY_INVALID', `invalid Ed25519 private key: ${e.message}`);
  }
}

function publicKeyObjectFromRaw(publicKeyHex) {
  if (!isLowercaseHex(publicKeyHex, RAW_PUBLIC_KEY_HEX_LENGTH)) {
    throw signatureError(
      'KDNA_INTEGRITY_SIGNATURE_FAILED',
      'signature bundle public_key must be a 32-byte Ed25519 key encoded as 64 lowercase hex characters',
    );
  }
  const der = Buffer.concat([SPKI_ED25519_PREFIX, Buffer.from(publicKeyHex, 'hex')]);
  try {
    return crypto.createPublicKey({ key: der, format: 'der', type: 'spki' });
  } catch (e) {
    throw signatureError(
      'KDNA_INTEGRITY_SIGNATURE_FAILED',
      `signature bundle public_key is not a valid Ed25519 key: ${e.message}`,
    );
  }
}

function rawPublicKeyHexFromObject(keyObject) {
  const spki = keyObject.export({ format: 'der', type: 'spki' });
  return spki.subarray(SPKI_ED25519_PREFIX.length).toString('hex');
}

/**
 * Generate a new Ed25519 signing key pair. The private key is returned as the
 * raw 32-byte seed in lowercase hex; custody is the signer's responsibility.
 */
function generateSigningKeyPair() {
  const { privateKey, publicKey } = crypto.generateKeyPairSync('ed25519');
  const pkcs8 = privateKey.export({ format: 'der', type: 'pkcs8' });
  const seedHex = pkcs8.subarray(PKCS8_ED25519_PREFIX.length).toString('hex');
  return {
    algorithm: KDSIG_ALGORITHM,
    private_key: seedHex,
    public_key: rawPublicKeyHexFromObject(publicKey),
  };
}

/**
 * Fingerprint of a raw Ed25519 public key: sha256 over the 32 key bytes.
 */
function keyFingerprint(publicKeyHex) {
  if (!isLowercaseHex(publicKeyHex, RAW_PUBLIC_KEY_HEX_LENGTH)) {
    throw signatureError(
      'KDNA_SIGNATURE_KEY_INVALID',
      'keyFingerprint expects a 32-byte Ed25519 public key encoded as 64 lowercase hex characters',
    );
  }
  return `sha256:${sha256Hex(Buffer.from(publicKeyHex, 'hex'))}`;
}

function stableStringify(value) {
  if (Array.isArray(value)) return `[${value.map(stableStringify).join(',')}]`;
  if (value && typeof value === 'object') {
    return `{${Object.keys(value)
      .sort()
      .map((key) => `${JSON.stringify(key)}:${stableStringify(value[key])}`)
      .join(',')}}`;
  }
  return JSON.stringify(value);
}

/**
 * Serialize a signature bundle to its canonical wire bytes (sorted keys, no
 * insignificant whitespace). Verification accepts any JSON encoding of the
 * same fields; canonical bytes are what signers and vectors pin.
 */
function serializeSignatureBundle(bundle) {
  return Buffer.from(`${stableStringify(bundle)}\n`, 'utf8');
}

function parseSignatureBundle(bytes) {
  let text;
  try {
    text = Buffer.isBuffer(bytes) ? bytes.toString('utf8') : String(bytes);
  } catch (e) {
    throw signatureError('KDNA_INTEGRITY_SIGNATURE_FAILED', `signature bundle is not UTF-8: ${e.message}`);
  }
  let bundle;
  try {
    bundle = JSON.parse(text);
  } catch (e) {
    throw signatureError(
      'KDNA_INTEGRITY_SIGNATURE_FAILED',
      `signature.kdsig is not valid JSON: ${e.message}`,
    );
  }
  if (!bundle || typeof bundle !== 'object' || Array.isArray(bundle)) {
    throw signatureError('KDNA_INTEGRITY_SIGNATURE_FAILED', 'signature bundle must be a JSON object');
  }
  for (const field of BUNDLE_FIELDS) {
    if (typeof bundle[field] !== 'string') {
      throw signatureError(
        'KDNA_INTEGRITY_SIGNATURE_FAILED',
        `signature bundle is missing required string field: ${field}`,
      );
    }
  }
  const extra = Object.keys(bundle).filter((key) => !BUNDLE_FIELDS.includes(key));
  if (extra.length) {
    throw signatureError(
      'KDNA_INTEGRITY_SIGNATURE_FAILED',
      `signature bundle carries unsupported fields: ${extra.sort().join(', ')}`,
    );
  }
  if (bundle.profile !== KDSIG_PROFILE) {
    throw signatureError(
      'KDNA_SIGNATURE_PROFILE_UNSUPPORTED',
      `signature bundle profile ${JSON.stringify(bundle.profile)} is not supported (supported: ${KDSIG_PROFILE})`,
    );
  }
  if (bundle.profile_version !== KDSIG_PROFILE_VERSION) {
    throw signatureError(
      'KDNA_SIGNATURE_VERSION_UNSUPPORTED',
      `signature bundle profile_version ${JSON.stringify(bundle.profile_version)} is not supported (supported: ${KDSIG_PROFILE_VERSION})`,
    );
  }
  if (bundle.algorithm !== KDSIG_ALGORITHM) {
    throw signatureError(
      'KDNA_SIGNATURE_PROFILE_UNSUPPORTED',
      `signature bundle algorithm ${JSON.stringify(bundle.algorithm)} is not supported (supported: ${KDSIG_ALGORITHM})`,
    );
  }
  assertContentDigestShape(bundle.content_digest);
  if (!isLowercaseHex(bundle.public_key, RAW_PUBLIC_KEY_HEX_LENGTH)) {
    throw signatureError(
      'KDNA_INTEGRITY_SIGNATURE_FAILED',
      'signature bundle public_key must be 64 lowercase hex characters',
    );
  }
  if (!isLowercaseHex(bundle.signature, RAW_SIGNATURE_HEX_LENGTH)) {
    throw signatureError(
      'KDNA_INTEGRITY_SIGNATURE_FAILED',
      'signature bundle signature must be 128 lowercase hex characters',
    );
  }
  return {
    profile: bundle.profile,
    profile_version: bundle.profile_version,
    algorithm: bundle.algorithm,
    content_digest: bundle.content_digest,
    public_key: bundle.public_key,
    signature: bundle.signature,
  };
}

/**
 * Sign a canonical content digest with an Ed25519 seed (64 lowercase hex).
 * Returns the bundle object; serialize with serializeSignatureBundle().
 */
function signContentDigest(contentDigest, privateKeySeedHex) {
  assertContentDigestShape(contentDigest);
  const privateKey = privateKeyObjectFromSeed(privateKeySeedHex);
  const publicKeyHex = rawPublicKeyHexFromObject(crypto.createPublicKey(privateKey));
  const signatureBytes = crypto.sign(null, buildSigningPayload(contentDigest), privateKey);
  return {
    algorithm: KDSIG_ALGORITHM,
    content_digest: contentDigest,
    profile: KDSIG_PROFILE,
    profile_version: KDSIG_PROFILE_VERSION,
    public_key: publicKeyHex,
    signature: signatureBytes.toString('hex'),
  };
}

/**
 * Verify a parsed bundle against a canonical content digest. Fail-closed:
 * every mismatch throws with a stable error code. Returns verification
 * evidence on success.
 */
function verifySignatureBundle(bundle, contentDigest, options = {}) {
  const parsed = parseSignatureBundle(bundle);
  assertContentDigestShape(contentDigest);
  if (parsed.content_digest !== contentDigest) {
    throw signatureError(
      'KDNA_INTEGRITY_SIGNATURE_FAILED',
      `signature bundle content_digest ${parsed.content_digest} does not match the asset content_digest ${contentDigest}`,
    );
  }
  if (
    typeof options.expectedPublicKey === 'string'
    && options.expectedPublicKey.toLowerCase() !== parsed.public_key
  ) {
    throw signatureError(
      'KDNA_INTEGRITY_SIGNATURE_FAILED',
      'signature bundle public_key does not match the expected pinned key',
    );
  }
  const publicKey = publicKeyObjectFromRaw(parsed.public_key);
  const signatureBytes = Buffer.from(parsed.signature, 'hex');
  let valid;
  try {
    valid = crypto.verify(null, buildSigningPayload(contentDigest), publicKey, signatureBytes);
  } catch (e) {
    throw signatureError(
      'KDNA_INTEGRITY_SIGNATURE_FAILED',
      `Ed25519 verification error: ${e.message}`,
    );
  }
  if (valid !== true) {
    throw signatureError(
      'KDNA_INTEGRITY_SIGNATURE_FAILED',
      'Ed25519 signature verification failed: signature does not match the canonical signing payload',
    );
  }
  return {
    state: 'verified',
    profile: parsed.profile,
    profile_version: parsed.profile_version,
    algorithm: parsed.algorithm,
    content_digest: parsed.content_digest,
    public_key: parsed.public_key,
    key_fingerprint: keyFingerprint(parsed.public_key),
  };
}

module.exports = {
  KDSIG_PROFILE,
  KDSIG_PROFILE_VERSION,
  KDSIG_ALGORITHM,
  SIGNATURE_ENTRY_NAME,
  buildSigningPayload,
  generateSigningKeyPair,
  keyFingerprint,
  parseSignatureBundle,
  serializeSignatureBundle,
  signContentDigest,
  verifySignatureBundle,
};

File: python-sdk/retired/kdna/core/signature.py
"""KDNA asset signatures — RFC-0021 M1 (``kdsig.ed25519``).

A ``.kdna`` container carries an optional top-level ``signature.kdsig`` entry:
a JSON signature bundle whose Ed25519 signature covers a deterministic
signing payload derived from the asset's canonical content digest
(``docs/CANONICALIZATION.md``). Verification is offline and fail-closed: any
malformed, unsupported, or unverifiable bundle rejects the asset.

A signature proves integrity and provenance only. It never proves
expertise, truthfulness, safety, or fitness for purpose.

This module is the independent Python verifier/signer counterpart of
``packages/kdna-core/src/signature.js``; both consume the same
known-answer vectors under ``conformance/signature/``.
"""

from __future__ import annotations

import hashlib
import json
from typing import Any, Dict, Optional

from cryptography.hazmat.primitives.asymmetric.ed25519 import (
    Ed25519PrivateKey,
    Ed25519PublicKey,
)
from cryptography.exceptions import InvalidSignature

KDSIG_PROFILE = "kdsig.ed25519"
KDSIG_PROFILE_VERSION = "0.1.0"
KDSIG_ALGORITHM = "ed25519"
SIGNATURE_ENTRY_NAME = "signature.kdsig"

_RAW_PUBLIC_KEY_HEX_LENGTH = 64
_RAW_SIGNATURE_HEX_LENGTH = 128
_SHA256_HEX_LENGTH = 64

_BUNDLE_FIELDS = (
    "algorithm",
    "content_digest",
    "profile",
    "profile_version",
    "public_key",
    "signature",
)


class KDNASignatureError(ValueError):
    """Fail-closed signature error carrying a stable KDNA error code."""

    def __init__(self, code: str, message: str) -> None:
        super().__init__(message)
        self.code = code


def _is_lowercase_hex(value: Any, length: int) -> bool:
    return (
        isinstance(value, str)
        and len(value) == length
        and all(ch in "0123456789abcdef" for ch in value)
    )


def _assert_content_digest_shape(value: Any) -> None:
    if not (
        isinstance(value, str)
        and value.startswith("sha256:")
        and _is_lowercase_hex(value[len("sha256:"):], _SHA256_HEX_LENGTH)
    ):
        raise KDNASignatureError(
            "KDNA_INTEGRITY_SIGNATURE_FAILED",
            'signature bundle content_digest must be "sha256:<64 lowercase hex>", '
            f"got {value!r}",
        )


def build_signing_payload(content_digest: str) -> bytes:
    """Build the exact bytes that Ed25519 signs.

    The payload is domain-separated by the profile coordinate and binds the
    asset's canonical content digest, so a signature can never be replayed
    across profiles, versions, or assets.
    """
    _assert_content_digest_shape(content_digest)
    return f"{KDSIG_PROFILE}:{KDSIG_PROFILE_VERSION}:{content_digest}".encode("utf-8")


def key_fingerprint(public_key_hex: str) -> str:
    """Fingerprint of a raw Ed25519 public key: sha256 over the 32 key bytes."""
    if not _is_lowercase_hex(public_key_hex, _RAW_PUBLIC_KEY_HEX_LENGTH):
        raise KDNASignatureError(
            "KDNA_SIGNATURE_KEY_INVALID",
            "key_fingerprint expects a 32-byte Ed25519 public key encoded as "
            "64 lowercase hex characters",
        )
    return f"sha256:{hashlib.sha256(bytes.fromhex(public_key_hex)).hexdigest()}"


def generate_signing_key_pair() -> Dict[str, str]:
    """Generate a new Ed25519 signing key pair.

    The private key is returned as the raw 32-byte seed in lowercase hex;
    custody is the signer's responsibility.
    """
    private_key = Ed25519PrivateKey.generate()
    public_key = private_key.public_key()
    return {
        "algorithm": KDSIG_ALGORITHM,
        "private_key": private_key.private_bytes_raw().hex(),
        "public_key": public_key.public_bytes_raw().hex(),
    }


def _stable_serialize(bundle: Dict[str, Any]) -> bytes:
    """Canonical wire bytes: sorted keys, no insignificant whitespace.

    Matches the JS ``stableStringify`` byte-for-byte for bundle content.
    """
    return (
        json.dumps(bundle, sort_keys=True, separators=(",", ":"), ensure_ascii=False)
        + "\n"
    ).encode("utf-8")


def serialize_signature_bundle(bundle: Dict[str, Any]) -> bytes:
    return _stable_serialize(bundle)


def parse_signature_bundle(data: bytes) -> Dict[str, str]:
    """Parse and strictly validate a signature bundle. Fail-closed."""
    try:
        text = data.decode("utf-8")
    except UnicodeDecodeError as error:
        raise KDNASignatureError(
            "KDNA_INTEGRITY_SIGNATURE_FAILED",
            f"signature bundle is not UTF-8: {error}",
        ) from error
    try:
        bundle = json.loads(text)
    except json.JSONDecodeError as error:
        raise KDNASignatureError(
            "KDNA_INTEGRITY_SIGNATURE_FAILED",
            f"signature.kdsig is not valid JSON: {error}",
        ) from error
    if not isinstance(bundle, dict):
        raise KDNASignatureError(
            "KDNA_INTEGRITY_SIGNATURE_FAILED",
            "signature bundle must be a JSON object",
        )
    for field in _BUNDLE_FIELDS:
        if not isinstance(bundle.get(field), str):
            raise KDNASignatureError(
                "KDNA_INTEGRITY_SIGNATURE_FAILED",
                f"signature bundle is missing required string field: {field}",
            )
    extra = sorted(key for key in bundle if key not in _BUNDLE_FIELDS)
    if extra:
        raise KDNASignatureError(
            "KDNA_INTEGRITY_SIGNATURE_FAILED",
            f"signature bundle carries unsupported fields: {', '.join(extra)}",
        )
    if bundle["profile"] != KDSIG_PROFILE:
        raise KDNASignatureError(
            "KDNA_SIGNATURE_PROFILE_UNSUPPORTED",
            f"signature bundle profile {bundle['profile']!r} is not supported "
            f"(supported: {KDSIG_PROFILE})",
        )
    if bundle["profile_version"] != KDSIG_PROFILE_VERSION:
        raise KDNASignatureError(
            "KDNA_SIGNATURE_VERSION_UNSUPPORTED",
            f"signature bundle profile_version {bundle['profile_version']!r} is "
            f"not supported (supported: {KDSIG_PROFILE_VERSION})",
        )
    if bundle["algorithm"] != KDSIG_ALGORITHM:
        raise KDNASignatureError(
            "KDNA_SIGNATURE_PROFILE_UNSUPPORTED",
            f"signature bundle algorithm {bundle['algorithm']!r} is not supported "
            f"(supported: {KDSIG_ALGORITHM})",
        )
    _assert_content_digest_shape(bundle["content_digest"])
    if not _is_lowercase_hex(bundle["public_key"], _RAW_PUBLIC_KEY_HEX_LENGTH):
        raise KDNASignatureError(
            "KDNA_INTEGRITY_SIGNATURE_FAILED",
            "signature bundle public_key must be 64 lowercase hex characters",
        )
    if not _is_lowercase_hex(bundle["signature"], _RAW_SIGNATURE_HEX_LENGTH):
        raise KDNASignatureError(
            "KDNA_INTEGRITY_SIGNATURE_FAILED",
            "signature bundle signature must be 128 lowercase hex characters",
        )
    return {field: bundle[field] for field in _BUNDLE_FIELDS}


def sign_content_digest(content_digest: str, private_key_seed_hex: str) -> Dict[str, str]:
    """Sign a canonical content digest with an Ed25519 seed (64 lowercase hex)."""
    _assert_content_digest_shape(content_digest)
    if not _is_lowercase_hex(private_key_seed_hex, _RAW_PUBLIC_KEY_HEX_LENGTH):
        raise KDNASignatureError(
            "KDNA_SIGNATURE_KEY_INVALID",
            "Ed25519 private key must be a 32-byte seed encoded as 64 lowercase "
            "hex characters",
        )
    try:
        private_key = Ed25519PrivateKey.from_private_bytes(
            bytes.fromhex(private_key_seed_hex)
        )
    except ValueError as error:
        raise KDNASignatureError(
            "KDNA_SIGNATURE_KEY_INVALID",
            f"invalid Ed25519 private key: {error}",
        ) from error
    public_key = private_key.public_key()
    signature = private_key.sign(build_signing_payload(content_digest))
    return {
        "algorithm": KDSIG_ALGORITHM,
        "content_digest": content_digest,
        "profile": KDSIG_PROFILE,
        "profile_version": KDSIG_PROFILE_VERSION,
        "public_key": public_key.public_bytes_raw().hex(),
        "signature": signature.hex(),
    }


def verify_signature_bundle(
    data: bytes,
    content_digest: str,
    expected_public_key: Optional[str] = None,
) -> Dict[str, str]:
    """Verify a bundle against a canonical content digest. Fail-closed.

    Returns the verification evidence on success; raises KDNASignatureError
    with a stable code on any failure.
    """
    parsed = parse_signature_bundle(data)
    _assert_content_digest_shape(content_digest)
    if parsed["content_digest"] != content_digest:
        raise KDNASignatureError(
            "KDNA_INTEGRITY_SIGNATURE_FAILED",
            f"signature bundle content_digest {parsed['content_digest']} does not "
            f"match the asset content_digest {content_digest}",
        )
    if (
        isinstance(expected_public_key, str)
        and expected_public_key.lower() != parsed["public_key"]
    ):
        raise KDNASignatureError(
            "KDNA_INTEGRITY_SIGNATURE_FAILED",
            "signature bundle public_key does not match the expected pinned key",
        )
    try:
        public_key = Ed25519PublicKey.from_public_bytes(
            bytes.fromhex(parsed["public_key"])
        )
    except ValueError as error:
        raise KDNASignatureError(
            "KDNA_INTEGRITY_SIGNATURE_FAILED",
            f"signature bundle public_key is not a valid Ed25519 key: {error}",
        ) from error
    try:
        public_key.verify(
            bytes.fromhex(parsed["signature"]),
            build_signing_payload(content_digest),
        )
    except InvalidSignature as error:
        raise KDNASignatureError(
            "KDNA_INTEGRITY_SIGNATURE_FAILED",
            "Ed25519 signature verification failed: signature does not match the "
            "canonical signing payload",
        ) from error
    except ValueError as error:
        raise KDNASignatureError(
            "KDNA_INTEGRITY_SIGNATURE_FAILED",
            f"Ed25519 verification error: {error}",
        ) from error
    return {
        "state": "verified",
        "profile": parsed["profile"],
        "profile_version": parsed["profile_version"],
        "algorithm": parsed["algorithm"],
        "content_digest": parsed["content_digest"],
        "public_key": parsed["public_key"],
        "key_fingerprint": key_fingerprint(parsed["public_key"]),
    }


def sign_container_bytes(data: bytes, private_key_seed_hex: str) -> Dict[str, Any]:
    """Sign a packaged container per RFC-0021 M1 (`kdsig.ed25519`).

    The container must pass every validation gate after any pre-existing
    signature entry is removed; signing never certifies a broken asset.
    Returns the signed container bytes, the bundle, and the signed content
    digest. Fail-closed throughout.
    """
    from . import container
    from .load import _compute_content_digest
    from .pack import pack_entries
    from .validate import run_validate

    layout = container.read_layout(data)
    unsigned_entries = {
        name: entry for name, entry in layout.entries.items() if name != SIGNATURE_ENTRY_NAME
    }
    unsigned_layout = container.Layout(
        kind=layout.kind,
        entries=unsigned_entries,
        manifest=layout.manifest,
        payload=layout.payload,
        mimetype=layout.mimetype,
    )
    validation = run_validate(unsigned_layout)
    if not validation["overall_valid"]:
        raise KDNASignatureError(
            "KDNA_SIGNATURE_INPUT_INVALID",
            "cannot sign: container does not pass validation "
            f"({'; '.join(validation['problems'])})",
        )
    content_digest = _compute_content_digest(unsigned_layout)
    bundle = sign_content_digest(content_digest, private_key_seed_hex)
    bundle_bytes = serialize_signature_bundle(bundle)
    signed_entries = {name: entry.data for name, entry in unsigned_entries.items()}
    signed_entries[SIGNATURE_ENTRY_NAME] = bundle_bytes
    return {
        "container_bytes": pack_entries(signed_entries),
        "bundle": bundle,
        "bundle_bytes": bundle_bytes,
        "content_digest": content_digest,
    }
